crates/ruff/src/rules/pylint/rules/cyclic_import.rs

```diff
@@ -0,0 +1,354 @@
+                    self.has_cycles_helper(&import.module, stack, cycles, fully_visited, level + 1);
+                    stack.pop();
+                }
+            }
+        }
+        fully_visited.insert(name.clone());
+    }
+}
+
+/// PLR0914
+pub fn cyclic_import(
+    path: &Path,
+    package: Option<&Path>,
+    imports: &ImportMap,
+    cycles: &mut FxHashMap<Arc<str>, FxHashSet<Vec<Arc<str>>>>,
+) -> Option<Vec<Diagnostic>> {
+    let Some(package) = package else {
+        return None;
+    };
+    let module_name = to_module_path(package, path).unwrap().join(".");
+    // if the module name isn't in the import map, it can't possibly have cycles
+    debug!("Checking module {module_name}");
+    let Some((module_name, _)) = imports.module_to_imports.get_key_value(&module_name as &str) else {
+        return None;
+    };
+    if let Some(existing_cycles) = cycles.get(module_name) {
+        if existing_cycles.is_empty() {
+            return None;
+        }
+        debug!("Existing cycles: {existing_cycles:#?}");
+        Some(
+            existing_cycles
+                .iter()
+                .map(|cycle| {
+                    Diagnostic::new(
+                        CyclicImport {
+                            cycle: cycle
+                                .iter()
+                                .map(std::string::ToString::to_string)
+                                .collect::<Vec<_>>()
+                                .join(" -> "),
+                        },
+                        imports.module_to_imports[module_name][0].as_ref().into(),
+                    )
+                })
+                .collect::<Vec<Diagnostic>>(),
+        )
+    } else {
+        let cyclic_import_checker = CyclicImportChecker {
+            imports: &imports.module_to_imports,
+        };
+        let VisitedAndCycles {
+            fully_visited: mut visited,
+            cycles: new_cycles,
+        } = cyclic_import_checker.has_cycles(module_name);
+        // we'll always have new visited stuff if we have
+        let mut out_vec: Vec<Diagnostic> = Vec::new();
+        if let Some(new_cycles) = new_cycles {
+            debug!("New cycles {new_cycles:#?}");
+            for new_cycle in &new_cycles {
+                if let [first, the_rest @ ..] = &new_cycle[..] {
+                    if first == module_name {
+                        out_vec.push(Diagnostic::new(
+                            CyclicImport {
+                                cycle: new_cycle
+                                    .iter()
+                                    .map(std::string::ToString::to_string)
+                                    .collect::<Vec<_>>()
+                                    .join(" -> "),
+                            },
+                            imports.module_to_imports[module_name]
+                                .iter()
+                                .find(|m| &(m.module) == the_rest.first().unwrap())
```

Comment:
Can we document (here and in other places where we call `unwrap`) why calling `unwrap` is safe or, even better, restructure the code so that calling `unwrap` is unnecessary?

Reply:
Sure, I think here though I should just be able to use `the_rest[0]`, would that be more idiomatic?